### PR TITLE
RUM-5083 refactor: Allow different sampling rates for each metric

### PR DESCRIPTION
**Pull request by ncreated**
### What and Why?

📦 This PR introduces the ability to sample each SDK metric using distinct sample rates.

Previously, [all metrics](https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/3534554673/Telemetry+Metric+Specs) shared the same [`metricTelemetrySampleRate`](https://github.com/DataDog/dd-sdk-ios/blob/691fc07fbc32267ec313cd3c3590f1155a9573ac/DatadogRUM/Sources/RUMConfiguration.swift#L273-L276) (15%). To sample batch metrics at different rates in `RUM-5083`, this refactoring is necessary.

### How?

A new property, `sampleRate`, has been added to the telemetry message that carries metric data:

```swift
public struct MetricTelemetry {
    public let name: String
    public let attributes: [String: Encodable]

    /// The sample rate for this metric, applied in addition to the telemetry sample rate.
    /// ...
    /// This sample rate is applied in the telemetry receiver, after the metric has been processed by the SDK core (tail-based sampling).
    public let sampleRate: Float
}

public enum TelemetryMessage {
    // ...
    case metric(MetricTelemetry)
}
```
The [`metricTelemetrySampleRate`](https://github.com/DataDog/dd-sdk-ios/blob/691fc07fbc32267ec313cd3c3590f1155a9573ac/DatadogRUM/Sources/RUMConfiguration.swift#L273-L276) has been removed, and now each metric is sampled individually using the metric-specific `sampleRate` (15% by default).

As before, this sample rate is applied in addition to the user-controlled [`telemetrySampleRate`](https://github.com/DataDog/dd-sdk-ios/blob/691fc07fbc32267ec313cd3c3590f1155a9573ac/DatadogRUM/Sources/RUMConfiguration.swift#L213-L218) (20% by default) in the RUM `TelemetryReceiver`. Both rates are applied after telemetry is passed over the `MessageBus` to RUM (tail-based sampling):

Screenshot 2024-08-23 at 09 46 44

#### Alternative Considered: Head-Based Telemetry Sampling

I also explored the potential benefits of switching to head-based sampling for the entire telemetry (not just metrics). By dropping telemetry samples earlier, we could reduce the cost of processing them over the `MessageBus`.

To evaluate this, I benchmarked Shopist iOS app, running one of its scenarios for 50s on an iPhone 13 Pro Max in "Release" configuration. Each time, I measured the number and timings of messages passed over the `MessageBus`:

Screenshot 2024-08-23 at 09 54 48

**Conclusion:** Head-based telemetry sampling doesn't offer significant gains. In this benchmark, telemetry accounted for the fewest type of messages sent over the message bus (~7%) and contributed only a negligible processing duration (~17ms) to the overall 50s run.

Instead, we should focus on the number of `context` and `baggage` messages exchanged over the bus which are more numerous and way heavier. The full results of this benchmark and further conclusions will be published in a separate document.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### CODE REVIEW

File: DatadogInternal/Sources/Telemetry/Telemetry.swift [hunk 1]
```diff
@@ -58,11 +58,32 @@ public struct ConfigurationTelemetry: Equatable {
     public let useWorkerUrl: Bool?
 }
 
+public struct MetricTelemetry {
+    /// The default sample rate for metric events, applied in addition to the telemetry sample rate (20% by default).
```
Comment (mariedm) [inline DatadogInternal/Sources/Telemetry/Telemetry.swift:62]: The default telemetry sample rate is 20% and the default metric sample rate is 15%, correct? The comment might be a bit confusing here.

Comment (mariedm) [inline DatadogInternal/Sources/Telemetry/Telemetry.swift:62]:
```suggestion
    /// The default sample rate for metric events (15%), applied in addition to the telemetry sample rate (20% by default).
```

File: DatadogInternal/Sources/Telemetry/Telemetry.swift [hunk 2]
```diff
@@ -75,20 +96,28 @@ public protocol Telemetry {
...
+    /// Unlike the telemetry sample rate and tail-based sampling in `stopMethodCalled()`, this sample rate is applied at the start of the method call timing.
+    /// This head-based sampling reduces the impact of processing high-frequency metrics in the SDK core, as most samples can be dropped
+    /// before being passed to the message bus.
+    ///
+    /// - Returns: A `MethodCalledTrace` instance for stopping the method call and measuring its execution time, or `nil` if the method call is not sampled.
     func startMethodCalled(
         operationName: String,
         callerClass: String,
-        samplingRate: Float = 100.0
+        sampleRate: Float
```
Comment (mariedm) [inline DatadogInternal/Sources/Telemetry/Telemetry.swift:118]: What is the reason behind removing the default value here?

File: DatadogInternal/Sources/Telemetry/Telemetry.swift [hunk 3]
```diff
@@ -98,15 +127,38 @@ public extension Telemetry {
...
     ///   - metric: The `MethodCalledTrace` instance.
-    ///   - isSuccessful: A flag indicating if the method call was successful.
-    func stopMethodCalled(_ metric: MethodCalledTrace?, isSuccessful: Bool = true) {
+    ///   - isSuccessful: A flag indicating whether the method call was successful.
+    ///   - sampleRate: The sample rate for **tail-based** sampling of the metric, applied after the metric is processed by the SDK core.
+    ///     Defaults to `MetricTelemetry.defaultSampleRate` (15%).
+    func stopMethodCalled(
+        _ metric: MethodCalledTrace?,
+        isSuccessful: Bool = true,
+        sampleRate: Float = MetricTelemetry.defaultSampleRate
```
Comment (mariedm) [inline DatadogInternal/Sources/Telemetry/Telemetry.swift:143]: Would it be clearer for customers to have this parameter named `tailBasedSampleRate` here, and `headBasedSampleRate` in `startMethodCalled`?

**Review by mariedm — COMMENTED** [review]
Awesome description! 🔥 I just have a couple of questions, but it looks great.

File: DatadogInternal/Sources/Telemetry/Telemetry.swift (hunk 1 above)
Comment (ncreated) [reply DatadogInternal/Sources/Telemetry/Telemetry.swift]: Yes, it's exactly like you said. And this is very fair call, I updated the comment as suggested 👍.

File: DatadogInternal/Sources/Telemetry/Telemetry.swift (hunk 2 above)
Comment (ncreated) [reply DatadogInternal/Sources/Telemetry/Telemetry.swift:118]: At first, I wanted compiler to guide me on all places where this is used 🙂, but after I was done I thought it's simply safer to not provide default for this particular call. The `startMethodCall()` is a high-frequency metric, defaulting it to 100% and overlooking it by mistake may bring large perf regression.

File: DatadogInternal/Sources/Telemetry/Telemetry.swift (hunk 3 above)
Comment (ncreated) [reply DatadogInternal/Sources/Telemetry/Telemetry.swift]: Oh, this was great idea 👍 - I also found `sampleRate` confusing in both `start()` and `stop()` but changing to `headSampleRate` and `tailSampleRate` makes total sense here. Done ✅.

BTW, this is `DatadogInternal` API, so not meant for customer use.

**Review by ganeshnj — APPROVED** [review]

**Pull request merged**